### PR TITLE
Adds chi squared module

### LINKED ISSUE #2: Add Baker-Cousins style log likelihood
Write a method to perform calculation of log likelihood, based on Baker-Cousins method of setting limits

Will form ```get_log_likelihood``` method in #4

### LINKED ISSUE #4: Add class for chi squared calculation
Base class for handling chi squared calculation

## Inputs
* 1D data histogram
* 1D montecarlo histogram
* Systematics that need to be accounted for

## Methods
* ```get_log_likelihood```
* ```get_penalty_term``` (accounts for systematics)
* ```get_chi_squared```

## Output
* Calculated chi squared for montecarlo compare to data

### LINKED ISSUE #5: Add method for calculating penalty term
Method to calculate the penalty term which is used in the chi squared calculation. This takes into account any systematics which are associated with the fit.

Will form the ```get_penalty_term``` method in #4

### LINKED ISSUE #6: Add method for calculating chi squared
Chi squared is calculated by summing the log likelihood and penalty term in quadrature.

Will form the ```get_chi_squared``` method in #4. The methods written in #2 and #5 will be called from within this method.

### PR DESCRIPTION
**Pull request by ashleyrback**
This is the class described in #4, and provides the functionality required
by #2, #5 and #6, although the structure is modified. As such this pull request
fixes #2, closes #4, resolves #5 and fixes #6.

The module contains three functions to calculate chi squared for a number given
number of observed events and number of expected events. The forms of chi
squared calculation included are Pearson's, Neyman's and the Poisson Likelihood
chi squared.

The module also adds the `ChiSquared` class that acts as a calculator for a
given `form` of chi squared calculation. It can also be initialised with
penalty term information via the keyword argument `penalty_term`. Or this can
be specified when calling the `get_chi_squared` method. The method takes two 1D
numpy arrays, representing the observed (or "data") spectrum and the expected
(or "MC") spectrum, as arguments and iterates over the array, calculating the
chi squared for each bin using the appropriate function, and adding a penalty
term if required. The method returns the total chi squared comparing the two
spectra.

### CODE REVIEW

File: echidna/core/chi_squared.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                chi_squared_sum += log_likelihood(n_observed, n_expected)
+        if (self._form == "poisson_likelihood"):
+            chi_squared_sum *= 2.0
+
+        # Add penalty term
+        if penalty_term_set:
+            chi_squared_sum += math.pow(parameter_value/sigma, 2)
+        return chi_squared_sum
+        
+def pearson_chi_squared(n_observed, n_expected):
```
Comment (pgjones) [inline echidna/core/chi_squared.py]: Internal method? Maybe prefix with `_`

File: echidna/core/chi_squared.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    Args:
+      n_observed (float): Number of observed events
+      n_expected (float): Number of expected events
+
+    Returns:
+      float. Calculated Pearson's chi squared
+    """
+    if (n_expected == 0.0):
+        raise ZeroDivisionError("expected events --> float divison by zero")
```
Comment (pgjones) [inline echidna/core/chi_squared.py]: This isn't dealt with anywhere, is that left to the user? Could the chi squared method not note this and reduce the degrees of freedom?

**Comment by pgjones** [comment]
Also generally did you consider something like:

    numpy.sum((x - y)**2/y)

as it might be much quicker?

File: echidna/core/chi_squared.py (hunk 1 above)
Comment (ashleyrback) [inline echidna/core/chi_squared.py]: They don't have to be internal, could they not be useful elsewhere in the code?

**Comment by ashleyrback** [comment]
I ran a little test with two versions of the `ChiSquared` class, the original version and one that uses the `numpy.sum` expressions to calculate the chi squared values. I tested the two classes with a short script that is set up in the same way as the unittest for the class, but just calculates Pearson's chi squared followed by the Poisson Likelihood chi squared, for both versions of the class.

I then timed the two versions. For 10 bins with 1000 entries it makes no difference as it takes less than a millisecond in both cases, but for 1000 bins and 1e6 entries the results are:

```
9083.52727273
9372.02412335

Initial method
==============
0:00:02.330
Elapsed time: 0:00:00.016


9083.52727273
9372.02412335

Numpy method
============
0:00:02.331
Elapsed time: 0:00:00.000
```
The chi squared values are the same in both cases, which means that the `numpy.sum` method is working as expected. Also the original method takes 16 ms to perform the two calculations, where as the `numpy.sum` method still takes less than a millisecond, so it is indeed faster.

**Comment by jwaterfield** [comment]
Following output from pep8 checker:

chi_squared.py:5:1: E302 expected 2 blank lines, found 1
chi_squared.py:7:1: W293 blank line contains whitespace
chi_squared.py:9:74: W291 trailing whitespace
chi_squared.py:38:80: E501 line too long (80 characters)
chi_squared.py:85:26: E261 at least two spaces before inline comment
chi_squared.py:102:79: E225 missing whitespace around operator

File: echidna/core/chi_squared.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    correct_bin_content, message = check_bin_content(observed)
+    if not correct_bin_content:
+        raise ValueError("chi_squared.pearson_chi_squared: observed " + message)
+    correct_bin_content, message = check_bin_content(expected)
+    if not correct_bin_content:
+        raise ValueError("chi_squared.pearson_chi_squared: expected " + message)
+    return numpy.sum((observed-expected)**2 / observed)
+
+def log_likelihood(observed, expected):
+    """ Calculates the log likelihood.
```
Comment (jwaterfield) [inline echidna/core/chi_squared.py]: State that its Baker-Cousins likelihood here

**Comment by jwaterfield** [comment]
pep8 output for test_chi_squared

test_chi_squared.py:7:1: E302 expected 2 blank lines, found 1
test_chi_squared.py:14:80: E501 line too long (86 characters)
test_chi_squared.py:62:27: E225 missing whitespace around operator
test_chi_squared.py:69:1: W293 blank line contains whitespace
test_chi_squared.py:75:77: W291 trailing whitespace
test_chi_squared.py:75:68: E201 whitespace after '{'

File: echidna/test/test_chi_squared.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        energy_high = 10.0
+        data_spectrum = numpy.ndarray(shape=(n_bins), dtype=float)
+        data_spectrum.fill(0)
+
+        # fill mock data spectrum from random gaussian
+        entries = 1000
+        mu = 5.0
+        sigma = 1.5
+        for entry in range(0, entries):
+            energy = sigma*numpy.random.randn() + mu
```
Comment (jwaterfield) [inline echidna/test/test_chi_squared.py]: could also use energy = numpy.random.normal(mu,sigma)

**Comment by jwaterfield** [comment]
Few lines of trailing whitespace. Happy to merge once changed
chi_squared.py:138:59: W291 trailing whitespace
chi_squared.py:167:59: W291 trailing whitespace
chi_squared.py:200:59: W291 trailing whitespace

**Pull request merged**